### DISCUSSION TITLE
web-tree-sitter node.equals() always returns true?

### DISCUSSION THREAD

**Discussion by joelspadin** [tree-sitter/tree-sitter#811] [Q&A] [answered]
I am trying to use web-tree-sitter in a VS Code extension to provide code completion for a custom language. I am able to parse code and get a valid tree, but I'm running into a problem that makes it difficult to analyze that tree. It appears that `node.equals(other)` always returns `true` no matter which two nodes I'm comparing.

For example:

```ts
const node = tree.rootNode.namedDescendantForPosition(cursorPos);
console.log(node.type, node.equals(tree.rootNode));
```

Which prints `true` no matter what type of node it finds at the cursor. What is the expected behavior here?

I found this code, https://github.com/tree-sitter/tree-sitter/blob/857a9ed07b983e2e6cff6dc851fcf9b37aec8e5a/lib/binding_web/binding.js#L260-L266 and I honestly have no idea what it is trying to check, but as far as I can tell the condition which would return false can never occur. Checking in my JS debugger, `node[i]` always evaluates to `0` for i = 0 and `undefined` for i = 1-4 for every node.

I also noticed there's an `id` property which isn't exposed in the TypeScript bindings but does seem to accurately identify nodes that are equal. Should the `equals` function be something more like this?

```js
equals(other) {
  return this.id === other.id;
}
```

**Comment by maxbrunsfeld** [comment, marked as answer]
I think you're right. This `equals` function never got updated after the code got refactored. If you're interested, you could open a PR that fixes it, and adds unit test coverage in [node-test.js](https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/test/node-test.js).

**Reply by joelspadin** [reply]
Cool. I'll make a PR sometime over the next few days probably.

**Reply by joelspadin** [reply]
I think I have a fix ready, but I'm following the instructions on https://tree-sitter.github.io/tree-sitter/contributing and `scripts/test-wasm` isn't working for me. All the tests fail with this error:

```
  1) Node
       "before all" hook in "Node":
     ReferenceError: loadWebAssemblyModule is not defined
      at /home/joel/projects/tree-sitter/lib/binding_web/tree-sitter.js:1:45700
      at async /home/joel/projects/tree-sitter/lib/binding_web/test/helper.js:10:15
```

Any idea how to fix that?

I'm running Ubuntu 20.04 inside WSL2 and building with emscripten 2.0.8.

**Reply by maxbrunsfeld** [reply]
What version of node is on your path?

**Reply by joelspadin** [reply]
I first tried it with 12.18.0, and when that didn't work I tried again with 14.15.1, which gave the same error.

**Reply by maxbrunsfeld** [reply]
Ok, this `loadWebAssemblyModule` error is due to a recent change in Emscripten: https://github.com/emscripten-core/emscripten/pull/12574#issuecomment-736091206. I'll be updating the Tree-sitter code to work with the latest version.

**Comment by razzeee** [comment]
Might be unrelated, but I learned that building via docker is way easier

**Reply by joelspadin** [reply]
Thanks. I removed emscripten from my environment and rebuilt everything (I assume it built using Docker but it's hard to tell) and it looks like it worked.

**Comment by joelspadin** [comment]
PR up: https://github.com/tree-sitter/tree-sitter/pull/819
